source/mainnet/smart-contracts/best-practices/development.rst

```diff
@@ -0,0 +1,203 @@
+Moreover, you should not make any assumptions about energy consumption, or expect that the execution succeeds.
+You contract should be able to correctly handle situations when the call to an external contract fails.
+
+General recommendations
+^^^^^^^^^^^^^^^^^^^^^^^
+
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
+- *Think about the contract state*. Do not assume that the contract state stays the same after an external call (see :ref:`best-practices-reentrancy` for details).
+- *Protect from denial-of-service (DoS) attacks*. Use the *Pull over Push* pattern: avoid sending funds back (*Push*) to an unknown address as part of some complex operation in your smart contract.
+  It opens doors to DoS attacks since the contract you call might fail for various reasons, blocking your functionality from succeeding.
+  Instead, create a separate entrypoint allowing users, which could be smart contracts, to request funds back (*Pull*).
+  Note that it is safe to transfer to user addresses, because on Concordium it is guaranteed not to execute any code.
+
+.. _best-practices-reentrancy:
+
+Reentrancy
+^^^^^^^^^^
+
+The *reentrant behavior* is not specific to smart contracts: it is a well known issue in the context of concurrency.
+A procedure can be interrupted in the middle of its execution, run again in *another* execution context, and then continue execution from the interruption point.
+In case of smart contracts, each external call interrupts the execution and hands over control to unknown code.
+Do not treat external contract invocations as regular method calls.
+Instead, think of them as sending a message and temporarily pausing execution of your contract.
+The receiving side has full control of what to do next and can choose to call your contract again while it is still in the "paused" state waiting for the external call to be completed.
+Once the external call is completed, you might find your contract in a completely different state.
+See an :ref:`example <reentracny-unit-testing>`, based on the DAO contract vulnerability of how reentrancy can be discovered using unit testing.
+
+- Avoid changing the state after an external call: use the *Checks-Effects-Interactions* pattern: validate data, update the contract state, make external calls.
+- If you need to perform some state changes after an external call use `invoke_contract_read_only <https://docs.rs/concordium-std/latest/concordium_std/trait.HasHost.html#method.invoke_contract_read_only>`_.
+  If the read-only invocation succeeds, it ensures that the state has not been changed after returning from the external call.
+  Using ``invoke_contract`` covers most of the cases where protection from reentrancy is required.
```

Comment:
Did you mean `invoke_contract_read_only` here ?

Reply:
Yes! Good catch, thanks! ;)